### PR TITLE
IR-207: Get endpoints for Azure Stack Cloud

### PR DESCRIPTION
**Pull request by dmage**
When the registry operator is installed on Azure Stack Hub, it should be able to use its endpoints.

As Azure Stack Hub [does not support the latest storage API](https://docs.microsoft.com/en-us/azure-stack/user/azure-stack-profiles-azure-resource-manager-versions?view=azs-2102), we should downgrade `github.com/Azure/azure-storage-blob-go` (which uses 2018-11-09) and use Microsoft.Storage 2019-06-01.

### CODE REVIEW

File: pkg/storage/azure/azure.go [hunk 1]
```diff
@@ -153,23 +153,32 @@ func (d *driver) accountExists(storageAccountsClient storage.AccountsClient, acc
...
+	params := &storage.AccountPropertiesCreateParameters{
+		AllowBlobPublicAccess: to.BoolPtr(false),
+		MinimumTLSVersion:     storage.TLS12,
+	}
+
+	if strings.EqualFold(cloudName, "AZURESTACKCLOUD") {
+		// It seems Azure Stack Hub does not support new API.
+		kind = storage.Storage
+		params = &storage.AccountPropertiesCreateParameters{}
+	}
```
Comment (dmage) [inline pkg/storage/azure/azure.go:169]: @patrickdillon does it make sense?

**Comment by dmage** [comment]
test platfrom flake
/retest

File: pkg/storage/azure/azure.go (hunk 1 above)
Comment (patrickdillon) [reply pkg/storage/azure/azure.go:169]: Well, I'm not sure. It seems like ASH supports 2019-06-01, and therefore should support `AllowBlobPublicAccess` and  `MinimumTLSVersion`. (I don't know about `storage.StorageV2` vs `storage.Storage`. )

Are you basing the removal of these two fields on the fact that the hybrid profile does not have these fields or have you tested and get an error when those fields are included? I would expect that if this code works, you should be able to include those two fields with ASH.

Hopefully this comment is not too unclear.

Comment (dmage) [reply pkg/storage/azure/azure.go:169]: I tried and it didn't work. It returned

> storage.AccountsClient#Create: Failure sending request: StatusCode=0 -- Original Error: Code="InvalidParameter" Message="Parameters are not valid in the request" Target="" Details=[{"code":"AccountPropertyCannotBeUpdated","message":"Property storageAccount.properties.allowBlobPublicAccess that cannot be updated for the storage account was specified in the request.","target":"storageAccount.properties.allowBlobPublicAccess"},{"code":"AccountPropertyCannotBeUpdated","message":"Property storageAccount.properties.minimumTlsVersion that cannot be updated for the storage account was specified in the request.","target":"storageAccount.properties.minimumTlsVersion"}]

Comment (patrickdillon) [reply pkg/storage/azure/azure.go:169]: Ok. I think this explains why the hybrid profile uses `2017-10-01`, which doesn't have these fields.

Yeah, with your explanation, this code makes sense to me.

**Comment by patrickdillon** [comment]
@jsafrane could you check this to see if this would be affected by openshift/installer#5138 when we remove credentials from the cloud config?

File: pkg/storage/azure/azure.go [hunk 2]
```diff
@@ -292,7 +301,7 @@ func (d *driver) storageAccountsClient(cfg *Azure, environment autorestazure.Env
 		storageAccountsClient.Authorizer = d.authorizer
 	} else {
 		clientCredentialsConfig := auth.NewClientCredentialsConfig(cfg.ClientID, cfg.ClientSecret, cfg.TenantID)
```
Comment (patrickdillon) [inline pkg/storage/azure/azure.go:303]: @jsafrane to my untrained eye it looks like this is reading the creds from the config which would be a problem with openshift/installer#5138

Comment (dmage) [reply pkg/storage/azure/azure.go:303]: these creds are from cloud-credential-operator, the operator requests them via https://github.com/openshift/cluster-image-registry-operator/blob/master/manifests/01-registry-credentials-request-azure.yaml

Comment (patrickdillon) [reply pkg/storage/azure/azure.go:303]: Perfect. Thank you.

**Comment by dmage** [comment]
/assign @ricardomaraschini @xiuwang @bmcelvee @sferich888 

Please review this PR and tag appropriately, we really want to have it in 4.9.

File: manifests/07-operator-ibm-cloud-managed.yaml [hunk 3]
```diff
@@ -37,6 +37,8 @@ spec:
           value: docker.io/openshift/origin-docker-registry:latest
         - name: IMAGE_PRUNER
           value: quay.io/openshift/origin-cli:v4.0
+        - name: AZURE_ENVIRONMENT_FILEPATH
+          value: /tmp/azurestackcloud.json
```
Comment (ricardomaraschini) [inline manifests/07-operator-ibm-cloud-managed.yaml:41]: Is this some kinda default location for this? I could not see where this file is used, IIUC the endpoints are copied into it but I could not find where it is used.

Comment (dmage) [reply manifests/07-operator-ibm-cloud-managed.yaml:41]: azure sdk uses AZURE_ENVIRONMENT_FILEPATH to find its config

Comment (ricardomaraschini) [reply manifests/07-operator-ibm-cloud-managed.yaml:41]: Oh yes, of course. Shame on me.

**Review by ricardomaraschini — APPROVED** [review]
/lgtm

**Comment by rpattath** [comment]
@dmage Is there an ETA on when this PR would be merged so that QE can be unblocked with functional testing on Azure Stack Hub? cc: @mgahagan73 @xiuwang

**Comment by dmage** [comment]
https://prow.ci.openshift.org/view/gs/origin-ci-test/pr-logs/pull/openshift_cluster-image-registry-operator/710/pull-ci-openshift-cluster-image-registry-operator-master-e2e-azure-operator/1428355090690871296
The test TestImageRegistryRemovedWithImages is known to be broken.

/override ci/prow/e2e-azure-operator

**Comment by dmage** [comment]
https://prow.ci.openshift.org/view/gs/origin-ci-test/pr-logs/pull/openshift_cluster-image-registry-operator/710/pull-ci-openshift-cluster-image-registry-operator-master-e2e-aws-operator/1430310582862483456
Same for AWS.

/override ci/prow/e2e-aws-operator

**Pull request merged**